queryable-spec.ts

```diff
@@ -0,0 +1,141 @@
+
+/*
+ * Helper union type 
+ */
+type termName = 'subject' | 'predicate' | 'object' | 'graph';
+
+/*
+ * Custom typings for the RDF/JS Stream interface as the current
+ * typings restrict the generic param Q to extensions of "BaseQuad",
+ * meaning it cannot be used for Bindings.
+ */
+export interface Stream<Q> extends EventEmitter {
+  read(): Q | null;
+}
+
+/*
+ * Map-like representation of Bindings as using plain objects could lead
+ * to collisions between variable names and object properties.
+ * Long-term goal: maintain compatibility with the native Map class.
+ */ 
+interface Bindings {
+  type: 'bindings';
+  get(variable: RDF.Variable): RDF.Term;
+  keys(): RDF.Variable[];
+  entries(): [RDF.Variable, RDF.Term][];
+  size: number;
+}
+
+/*
+ * Bindings objects are created using a dedicated factory, keeping in line
+ * with DataFactory.quad(). This also helps with facilitating support for
+ * immutability. Basic helper methods must also be provided for the most 
+ * common manipulations of bindings objects.
+ */
+interface BindingsFactory {
+  bindings(entries: [RDF.Variable, RDF.Term][]): Bindings;
+  merge(bindings: Bindings[]): Bindings;
+}
+
+/*
+ * Base interfaces to represent query results. These interfaces are generic
+ * with respect to the types of query metadata objects. These can be extended
+ * by implementors.
+ */
+
+interface BaseQueryResultMetadata<OrderItemsType> {
+  cardinality?: number; // Cardinality estimate
+  order?: OrderItemsType[];
+}
+
+interface BaseQueryResult<MetadataType extends BaseQueryResultMetadata<RDF.Variable | termName>> {
+  type: 'bindings' | 'quads' | 'boolean';
+  metadata(opts: Partial<Record<keyof MetadataType, boolean>>): Promise<MetadataType>;
+}
+
+interface QueryResultBindings<MetadataType extends BaseQueryResultMetadata<RDF.Variable>> extends BaseQueryResult<MetadataType> {
+  type: 'bindings';
+  bindings(): Promise<Bindings[]>;
+  stream(): Stream<Bindings>;
+  variables: RDF.Variable[];
+}
+    
+interface QueryResultQuads<MetadataType extends BaseQueryResultMetadata<termName>> extends BaseQueryResult<MetadataType> {
+  type: 'quads';
+  quads(): Promise<RDF.Quad[]>;
+  stream(): Stream<RDF.Quad>;
+}
+
+interface QueryResultBoolean<MetadataType extends BaseQueryResultMetadata<any>> extends BaseQueryResult<MetadataType> {
+  type: 'boolean';
+  value: Promise<boolean>;
+}
+
+type QueryResult<M> = QueryResultBindings<M> | QueryResultQuads<M> | QueryResultBoolean<M>;
+
+/*
+ * Context objects provide a way to pass additional bits information to
+ * implementors, such as but not limited to:
+ * - data sources
+ * - base IRI for IRI resolution
+ * - timestamp for expression evaluation
+ * - query language
+ * - ...
+ */
+
+// SourceType can be anything the query engine defines
+// TODO: we may consider defining some standards, like 'string', RDF.Source, ...
+interface QueryContext<SourceType> {
+  sources: [SourceType, ...SourceType[]];
```

Comment:
I would make this optional, as certain engines may decide to pass the source differently (e.g. via the constructor).

Reply:
Curious to hear what others think of this, e.g. @tpluscode and @bergos.

Reply:
This has been open for a while. In the absence of further comments, I would go for what I think is the most flexible option, i.e. the current approach of defining sources in the query context.

Reply:
No activity on this one, so I'll consider this resolved with the current approach. Happy to revisit based on new feedback.